jdaviz/configs/specviz/plugins/viewers.py

```diff
@@ -383,84 +372,79 @@ def add_data(self, data, color=None, alpha=None, **layer_state):
         return result
 
     def _plot_mask(self):
+        if not self.display_mask:
+            return
 
-        _data = self._data[0]
-        _data_trace = self.data_trace_pointer
-
-        if 'mask' in _data.components and self.display_mask:
-
-            # If a mask was already plotted, remove it.
-            if (hasattr(self, 'mask_trace_pointer') and
-                    self.mask_trace_pointer is not None):
-                marks = self.figure.marks[:]
-                self._clean_mask(marks)
-                self.figure.marks = marks
-
-            mask = _data['mask'].data
-
-            # get trace with spectrum.
-            x = self.figure.marks[_data_trace].x
-            y = self.figure.marks[_data_trace].y
-
-            # For plotting markers only for the masked data
-            # points, erase un-masked data from trace.
-            y = np.where(np.asarray(mask) == 0, np.nan, y)
-
-            # there is no 'X' marker option in bqplot
-            self.mask_line_mark = Scatter(scales=self.scales,
-                                          marker='cross',
-                                          x=x,
-                                          y=y,
-                                          stroke_width=0.5,
-                                          # colors=['red'],
-                                          colors=[self._color],
-                                          default_size=25,
-                                          default_opacities=[self._alpha],
-                                          )
-            self.figure.marks = list(self.figure.marks) + [self.mask_line_mark]
-
-            # We added an extra trace. Get pointer to it.
-            self.mask_trace_pointer = len(self.figure.marks) - 1
-
+        # Loop through all active data in the viewer
+        for index, layer_state in enumerate(self.state.layers):
+            comps = [str(component) for component in layer_state.layer.components]
+
+            # Ignore data that does not have an 'mask' component
+            if "mask" in comps:
+                mask = np.array(layer_state.layer['mask'].data)
+
+                data_x = layer_state.layer.data.get_object().spectral_axis
+                data_y = layer_state.layer.data.get_object().flux.value
+
+                # For plotting markers only for the masked data
+                # points, erase un-masked data from trace.
+                y = np.where(np.asarray(mask) == 0, np.nan, data_y)
+
+                # A subclass of the bqplot Scatter object, ScatterMask places
+                # 'X' marks where there is masked data in the viewer.
```

Comment:
ok, that could be handy!  Is this controlled by the same switch as uncertainties or should they be independent?